<a href="https://colab.research.google.com/github/gilesitorr/DataScience3_Bloque3/blob/main/Postwork_5/Readme.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Postwork 05: Limpieza de datos y agregaciones



1.   Limpiar nuestro dataset de NaNs.
2.   Reindexar, si es necesario.
3.   Renombrar columnas si es necesario.
4.   Experimentar la aplicación de agregaciones para explorar nuestro dataset.



In [ ]:
#Se cuentan los NaNs en cada columna
df.isna().sum(axis=0)

Date                              0
Time                              0
Latitude                          0
Longitude                         0
Type                              0
Depth                             0
Depth Error                   18951
Depth Seismic Stations        16315
Magnitude                         0
Magnitude Type                    3
Magnitude Error               23085
Magnitude Seismic Stations    20848
Azimuthal Gap                 16113
Horizontal Distance           21808
Horizontal Error              22256
Root Mean Square               6060
ID                                0
Source                            0
Location Source                   0
Magnitude Source                  0
Status                            0
dtype: int64

In [ ]:
valores_nulos_porcentaje=df.isna().mean()
valores_nulos_porcentaje.map(lambda x: str(round(x*100,2))+' %')

Date                            0.0 %
Time                            0.0 %
Latitude                        0.0 %
Longitude                       0.0 %
Type                            0.0 %
Depth                           0.0 %
Depth Error                   80.95 %
Depth Seismic Stations        69.69 %
Magnitude                       0.0 %
Magnitude Type                 0.01 %
Magnitude Error                98.6 %
Magnitude Seismic Stations    89.05 %
Azimuthal Gap                 68.82 %
Horizontal Distance           93.15 %
Horizontal Error              95.06 %
Root Mean Square              25.88 %
ID                              0.0 %
Source                          0.0 %
Location Source                 0.0 %
Magnitude Source                0.0 %
Status                          0.0 %
dtype: object

Borraremos las columnas "Depth Error", "Depth Seismic Stations", "Magnitude Error", "Magnitude Seismic Stations", "Azimuthal Gap", "Horizontal Distance", "Horizontal Error" ,"Root Mean Square" debido a la cantidad de NaNs encontrados en ellas.     

In [ ]:
df_dropped=df.dropna(axis=1, thresh=23400)

In [ ]:
df_dropped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23412 entries, 0 to 23411
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              23412 non-null  object 
 1   Time              23412 non-null  object 
 2   Latitude          23412 non-null  float64
 3   Longitude         23412 non-null  float64
 4   Type              23412 non-null  object 
 5   Depth             23412 non-null  float64
 6   Magnitude         23412 non-null  float64
 7   Magnitude Type    23409 non-null  object 
 8   ID                23412 non-null  object 
 9   Source            23412 non-null  object 
 10  Location Source   23412 non-null  object 
 11  Magnitude Source  23412 non-null  object 
 12  Status            23412 non-null  object 
dtypes: float64(4), object(9)
memory usage: 2.3+ MB


In [ ]:
df_dropped.isna().sum()

Date                0
Time                0
Latitude            0
Longitude           0
Type                0
Depth               0
Magnitude           0
Magnitude Type      3
ID                  0
Source              0
Location Source     0
Magnitude Source    0
Status              0
dtype: int64

Como en la columna "Magnitude Type" hay tres valores NaN's, se eliminarán esas tres filas.

In [ ]:
df_no_nan = df_dropped.dropna()
df_no_nan

Date      Time  ...  Magnitude Source     Status
0      01/02/1965  13:44:18  ...            ISCGEM  Automatic
1      01/04/1965  11:29:49  ...            ISCGEM  Automatic
2      01/05/1965  18:05:58  ...            ISCGEM  Automatic
3      01/08/1965  18:49:43  ...            ISCGEM  Automatic
4      01/09/1965  13:32:50  ...            ISCGEM  Automatic
...           ...       ...  ...               ...        ...
23407  12/28/2016  08:22:12  ...                NN   Reviewed
23408  12/28/2016  09:13:47  ...                NN   Reviewed
23409  12/28/2016  12:38:51  ...                US   Reviewed
23410  12/29/2016  22:30:19  ...                US   Reviewed
23411  12/30/2016  20:08:28  ...                US   Reviewed

[23409 rows x 13 columns]

Se comprueba que no exista ningún NaN en el DataFrame.



In [ ]:
df_no_nan.isna().sum()

Date                0
Time                0
Latitude            0
Longitude           0
Type                0
Depth               0
Magnitude           0
Magnitude Type      0
ID                  0
Source              0
Location Source     0
Magnitude Source    0
Status              0
dtype: int64

In [ ]:
df_no_nan

Date      Time  ...  Magnitude Source     Status
0      01/02/1965  13:44:18  ...            ISCGEM  Automatic
1      01/04/1965  11:29:49  ...            ISCGEM  Automatic
2      01/05/1965  18:05:58  ...            ISCGEM  Automatic
3      01/08/1965  18:49:43  ...            ISCGEM  Automatic
4      01/09/1965  13:32:50  ...            ISCGEM  Automatic
...           ...       ...  ...               ...        ...
23407  12/28/2016  08:22:12  ...                NN   Reviewed
23408  12/28/2016  09:13:47  ...                NN   Reviewed
23409  12/28/2016  12:38:51  ...                US   Reviewed
23410  12/29/2016  22:30:19  ...                US   Reviewed
23411  12/30/2016  20:08:28  ...                US   Reviewed

[23409 rows x 13 columns]

Se reindexa el DataFrame para terminar la limpieza

In [ ]:
#Como se hará un reindexado, la columna "index" es inútil, por lo que se le hará drop
#para que no se cree una columna con los índices antiguos
df_clean = df_no_nan.reset_index(drop=True)
df_clean

Date      Time  ...  Magnitude Source     Status
0      01/02/1965  13:44:18  ...            ISCGEM  Automatic
1      01/04/1965  11:29:49  ...            ISCGEM  Automatic
2      01/05/1965  18:05:58  ...            ISCGEM  Automatic
3      01/08/1965  18:49:43  ...            ISCGEM  Automatic
4      01/09/1965  13:32:50  ...            ISCGEM  Automatic
...           ...       ...  ...               ...        ...
23404  12/28/2016  08:22:12  ...                NN   Reviewed
23405  12/28/2016  09:13:47  ...                NN   Reviewed
23406  12/28/2016  12:38:51  ...                US   Reviewed
23407  12/29/2016  22:30:19  ...                US   Reviewed
23408  12/30/2016  20:08:28  ...                US   Reviewed

[23409 rows x 13 columns]

Para tener una idea burda de la región geográfica que aparece más en el dataset, se calcula el promedio y la desviación estándar de la longitud y de la latitud

In [ ]:
#Latitud
print(f"La distribución de latitud es\n{df_clean['Latitude'].mean():.2f} +/- {df_clean['Latitude'].std():.2f}")

La distribución de latitud es
1.67 +/- 30.11


In [ ]:
#Longitud
print(f"La distribución de longitud es\n{df_clean['Longitude'].mean():.2f} +/- {df_clean['Longitude'].std():.2f}")

La distribución de longitud es
39.66 +/- 125.51


Para poner en perspectiva la distribución anterior, se debe conocer el intervalo de valores de ambas columnas

In [ ]:
#Latitud
print(f"La latitud toma valores entre\n{df_clean['Latitude'].min():.2f} y {df_clean['Latitude'].max():.2f}")

La latitud toma valores entre
-77.08 y 86.00


In [ ]:
#Longitud
print(f"La longitud toma valores entre\n{df_clean['Longitude'].min():.2f} y {df_clean['Longitude'].std():.2f}")

La longitud toma valores entre
-180.00 y 125.51


Concluimos que la distribución en la latitud no es tan dispersa, pero la distribución sobre la longitud sí.

Para tener una perspectiva condensada de lo anterior, se usa la agregación _describe_

Se confirma lo que se  mencionó sobre la distribución ya no sólo con el promedio y la desviación estándar, sino también con los cuartiles

In [ ]:
df_clean.describe()

Latitude     Longitude         Depth     Magnitude
count  23409.000000  23409.000000  23409.000000  23409.000000
mean       1.674322     39.660642     70.775695      5.882553
std       30.112233    125.506701    122.657829      0.423087
min      -77.080000   -179.997000     -1.100000      5.500000
25%      -18.656000    -76.311300     14.540000      5.600000
50%       -3.570000    103.993000     33.000000      5.700000
75%       26.158000    145.027000     54.000000      6.000000
max       86.005000    179.998000    700.000000      9.100000

Por último, puede contarse la cantidad de registros de cada tipo. 

In [ ]:
df_clean['Type'].value_counts()

Earthquake           23229
Nuclear Explosion      175
Explosion                4
Rock Burst               1
Name: Type, dtype: int64

De los 23,409 registros, 23,229 son terremotos, por lo que la base de datos es útil. 